### DISCUSSION TITLE
AVX2 parallelization

### DISCUSSION THREAD

**Discussion by bulatkhusainov** [casadi/casadi#3744] [Q&A]
I am using Casadi for real-time MPC and my dynamics function contains a number of `for` loops, where a function is evaluated multiple times for different data. The intention is to parallelize these `for` loops with AVX2 functionality of the target CPU. However, I did not observe any speed up when applying /O2 /arch:AVX2 /fp:fast compiler options in Visual Studio 2019.

Consider a small MATLAB example to illustrate the problem:
```
x = MX.sym('x');
f = Function('f',{x}, {sin(x)});
fVect = f.map(2,'serial');
fVect.generate('gen.c');
```

I can think of the following reasons why the compiler may decide not to vectorise the `for` loop at the bottom of `casadi_f0()` funciton.

- The loop contains `if` statements
- ` casadi_f1()` function body contains `if` statements
- There may be loop dependencies, e.g. due to `arg1[0]+=1;` 
- Related to above: `casadi_f1()` may be not unlined.

I understand from [this paper](https://lirias.kuleuven.be/retrieve/640748) that SIMD parallelization is possible with Casadi, but clearly I am missing something.

I would appreciate any suggestions on SIMD parallelization with Casadi.

Generated C code:
```
/* f:(i0)->(o0) */
static int casadi_f1(const casadi_real** arg, casadi_real** res, casadi_int* iw, casadi_real* w, int mem) {
  casadi_real w0;
  /* #0: @0 = input[0][0] */
  w0 = arg[0] ? arg[0][0] : 0;
  /* #1: @0 = sin(@0) */
  w0 = sin( w0 );
  /* #2: output[0][0] = @0 */
  if (res[0]) res[0][0] = w0;
  return 0;
}

/* map2_f:(i0[1x2])->(o0[1x2]) */
static int casadi_f0(const casadi_real** arg, casadi_real** res, casadi_int* iw, casadi_real* w, int mem) {
  casadi_int i;
  casadi_real* *res1;
  const casadi_real* *arg1;
  arg1 = arg+1;
  for (i=0; i<1; ++i) arg1[i]=arg[i];
  res1 = res+1;
  for (i=0; i<1; ++i) res1[i]=res[i];
  for (i=0; i<2; ++i) {
    if (casadi_f1(arg1, res1, iw, w, 0)) return 1;
    if (arg1[0]) arg1[0]+=1;
    if (res1[0]) res1[0]+=1;
  }
  return 0;
}
```

**Comment by jgillis** [comment]
Hi @bulatkhusainov 

I've had a very successful collaboration the past two years to achieve almost perfect auto-vectorization of a complex CasADi model.
Unfortunately, some of the involved strategies were specific to the application and some generic parts do not yet meet the coding standards needed for merging to main.
